trie-db/src/iterator.rs

```diff
@@ -56,31 +56,64 @@ impl<H: Hasher> Crumb<H> {
 			_ => Status::Exiting,
 		}
 	}
+
+	/// Move on to next status in the node's sequence in backward direction.
+	fn decrement(&mut self) {
+		self.status = match (self.status, self.node.node_plan()) {
+			(Status::Entering, NodePlan::Extension { .. }) => Status::At,
+			(Status::Entering, NodePlan::Branch { .. }) |
+			(Status::Entering, NodePlan::NibbledBranch { .. }) => Status::At,
+			(Status::At, NodePlan::Branch { .. }) |
+			(Status::At, NodePlan::NibbledBranch { .. }) => Status::AtChild(15),
+			(Status::AtChild(x), NodePlan::Branch { .. }) |
+			(Status::AtChild(x), NodePlan::NibbledBranch { .. })
+				if x > 0 =>
+				Status::AtChild(x - 1),
+			_ => Status::Exiting,
+		}
+	}
 }
 
 /// Iterator for going through all nodes in the trie in pre-order traversal order.
 pub struct TrieDBRawIterator<L: TrieLayout> {
+	/// Forward trail of nodes to visit.
 	trail: Vec<Crumb<L::Hash>>,
+	/// Forward iteration key nibbles of the current node.
 	key_nibbles: NibbleVec,
+	/// Backward trail of nodes to visit.
+	back_trail: Vec<Crumb<L::Hash>>,
+	/// Backward iteration key nibbles of the current node.
+	back_key_nibbles: NibbleVec,
```

Comment:
Did you try simply using `trail` and `key_nibbles` for backward iteration? Not duplicating it would be good as it would make switching between next and previous call simple.
Also may remove the need to implement clone.

Reply:
How would you track both cursors iterating forward and backwards? 

The `next_raw_item` takes in `&mut self` and passing a direction would not be enough to simutaneously iterate in both directions. Are you implying that we should add a `back_raw_iter` to every iterator struct like this?

```
/// Iterator for going through all nodes in the trie in pre-order traversal order.
pub struct TrieDBNodeIterator<'a, 'cache, L: TrieLayout> {
	db: &'a TrieDB<'a, 'cache, L>,
	raw_iter: TrieDBRawIterator<L>,
	back_raw_iter: TrieDBRawIterator<L>,
}
```

Or you think there should be a single cursor going backwards and forwards? (I think this would not be ideal, since you would need to iterate much more if you wanted to fetch 2 nodes simutaneously).

Reply:
> Or you think there should be a single cursor going backwards and forwards? 

Yes, double end iterator should be able to do both.

> (I think this would not be ideal, since you would need to iterate much more if you wanted to fetch 2 nodes simutaneously).

If you need two nodes, then you would need two instance of iterator or am I missing something?

Reply:
> If you need two nodes, then you would need two instance of iterator or am I missing something?

Ok I see, so you think we should have a single cursor. I did not go for this initially since I was basing my implementation on the [`DoubleEndedIterators`](https://doc.rust-lang.org/std/iter/trait.DoubleEndedIterator.html#examples) documentation which demonstrates 2 cursors. 

In that case, I am for using a single cursor cause it would simplify the implementation.

Reply:
Oh yes you are right, the actual trait is working on two pointers.
For trie crate it means doing what you propose:
```
// Iterator for going through all nodes in the trie in pre-order traversal order.
pub struct TrieDBNodeIterator<'a, 'cache, L: TrieLayout> {
	db: &'a TrieDB<'a, 'cache, L>,
	raw_iter: TrieDBRawIterator<L>,
	back_raw_iter: TrieDBRawIterator<L>,
}
```
would be better to be a different iterator (eg DoubleTrieDBNodeIterator), to keep the smaller struct when doubleended is not needed (keeping `next_back_item` public for it make sense to me, so we can still use the standard iterator in both direction (the use case of DoubleTrieDBNodeIterator is a bit off what I would expect).

Reply:
> (keeping `next_back_item` public for it make sense to me, so we can still use the standard iterator in both direction (the use case of DoubleTrieDBNodeIterator is a bit off what I would expect)

Not sure what you mean by this. Do you mean that you would still be able to iterate with a single cursor if using the raw iterator directly?

Reply:
keeping TrieDBNodeIterator as is and having a new struct (as you describe) :
```
pub struct TrieDBNodeDoubleEndedIterator<'a, 'cache, L: TrieLayout> {
	db: &'a TrieDB<'a, 'cache, L>,
	raw_iter: TrieDBRawIterator<L>,
	back_raw_iter: TrieDBRawIterator<L>,
}
```
and a new function to get it (eg `fn double_ended_iter(self: &TrieDB) -> TrieDBNodeDoubleEndedIterator`)

Reply:
I don't think we should allow iterating backwards with any of the standard iterators. To iterate backwards, they would have to use `TrieDBNodeDoubleEndedIterator`. 

The api is a little confusing having the possibility to iterate backwards with a single cursor for a regular iterator. There should be an explicit distinction between these two types.

Also iterating backwards with a single cursor can result in ending the iteration if it reaches the back of the trie. To me this is even more confusing, since the original implementation relies on reaching the end of the trie (eg the last node in the trie) to end. It also adds more complexity.

Reply:
The change for iterating backward looks simple enough to me ( https://github.com/snowmead/trie/compare/master...cheme:trie:208 )
But I agree api gets a bit confusing when it comes to have a specific doubleendediterator struct.
But looking at my changes, it seems easy enough to just not use the trait (see change in the test).
I mean in our case the double ended nature is not part of our iterator and requires doubling the struct size.

> Also iterating backwards with a single cursor can result in ending the iteration if it reaches the back of the trie. 

Yes that s a good point, not sure how problematic it can be though.
Can be fix by always keeping root node or prefix node in the crumb list, but I am not sure the change is worth it.

Maybe we just want to expose a backward iterator (IIRC some std behavior for `.rev`  did surprise me a bit in the past, but as long as it is call before first iteration it is fairly ok).

Reply:
Answered outside of this thread, since it impacts more then the initial comment. https://github.com/paritytech/trie/pull/208#issuecomment-1871442289